### ISSUE TITLE
Reflect don't recognize foreign tables in postgresql

### ISSUE THREAD

**Issue by joaostorrer** [sqlalchemy/sqlalchemy#9944]
### Describe the use case

Currently, when I call the MetaData().reflect() function, the foreign tables are not reflected. It would be interesting to have an argument in the function to be able to search for foreign tables as well.
This way facilitates DML operations on foreign tables.

### Databases / Backends / Drivers targeted

Database: PostgreSQL
Drivers: psycopg2

### Example Use

For simplicity, to connect one postgresql server to another (using foreign-data wrapper):

PostgreSQL 1:
- create database test;
- create role test login password 'test123';
- create table account (username varchar (50) unique not null, password varchar (50) not null;
- insert into account ('test', 'test123');

PostgreSQL 2:
- create database test_fdw;
- create role test_fdw login password 'testfdw123';
- create extension postgres_fdw;
- create server test foreign data wrapper postgres_fdw options (host 'PostgreSQL1', dbname 'test', port '5432');
- create user mapping server for test_fdw test options (user 'test', password 'test123');
- import foreign schema test from server test into test_fdw; (table account will be imported into test_fdw)

SQLAlchemy:
engine = create_engine("postgresql+psycopg2://test_fdw:testfdw123@PostgreSQL2:5432/test_fdw")
metadata_obj = MetaData()
metadata_obj.reflect(bind=engine)
account = metadata_obj.tables["account"]

**Comment by zzzeek** [comment]
ok ummmm

is there an information schema we can access?    the docs at https://www.postgresql.org/docs/current/postgres-fdw.html assuming that's what we're talking about are quite vague.  like what does [this](https://www.postgresql.org/docs/current/sql-importforeignschema.html) do?   

if there's not any information schema then this is a non-starter; just log into that other database directly and reflect into the same metadata object with a schema prefix added.   also if there's anything more involved on the reflection side than "look at information schema from this alternate schema", it's probably not worth doing.

note that when you say "interesting", we have... a lot of "interesting" things to work on already, so at the same time, if this is just "interesting" and not a mission critical use case, I'd rather pass on it, but let me know some more detail how exactly we would access these schemas via the fdw interface, as opposed to simply making another engine that logs into that DB directly, in which case we will gladly give you a recipe to do this and put it on the wiki.

**Comment by joaostorrer** [comment]
As foreign tables aparecem nos information schemas "foreign_tables" e "tables" e as colunas das foreign tables podem ser acessadas no information schema "columns".
A função "import foreign schema" apenas busca todos os objetos do outro servidor postgres e cria as foreign tables correspondentes no servidor postgres atual, ou seja, é apenas uma facilidade do postgres para que não seja necessário criar manualmente cada foreign table.
Os foreign data wrappers (fdw) do postgres são mais utilizados para integrar vários conjuntos de dados no postgres, de tal forma que é possível realizar queries entre vários conjuntos de dados diferentes de forma simples.
No projeto que eu estou trabalhando atualmente eu utilizo as foreign tables nesse sentido, mas eu também preciso inserir e atualizar dados em foreign tables de diferentes servidores.

**Comment by zzzeek** [comment]
yes if they are in the local information schema then this would already work

**Comment by joaostorrer** [comment]
I think i found the problem, the function get_foreign_table_names was created, but is not used anywhere:

https://github.com/sqlalchemy/sqlalchemy/blob/21b2661660132c4258aad18cade78ce0e0539688/lib/sqlalchemy/dialects/postgresql/base.py#L2795-L2809

reflect only calls get_table_names, get_view_names and get_materialized_view_names functions:

https://github.com/sqlalchemy/sqlalchemy/blob/21b2661660132c4258aad18cade78ce0e0539688/lib/sqlalchemy/sql/schema.py#L5704-L5713

**Comment by zzzeek** [comment]
@CaselIT im thinking we might want to make some new API that supersedes `metadata.reflect()` and uses the `Inspector` directly, so that people can take advantage of what the "_get_reflection_info()" private method does now, that is, some way of getting an inspector and loading it up with table data, then populating a `MetaData` with `Table` objects from that.  this would also suit the "pickle" use case we saw the other day.

**Comment by CaselIT** [comment]
> @CaselIT im thinking we might want to make some new API that supersedes `metadata.reflect()` and uses the `Inspector` directly, so that people can take advantage of what the "_get_reflection_info()" private method does now, that is, some way of getting an inspector and loading it up with table data, then populating a `MetaData` with `Table` objects from that. this would also suit the "pickle" use case we saw the other day.

It may make sense.

The same issue is also on the temp tables btw.

**Comment by CaselIT** [comment]
> OK then you can reflect these semi-manually like this:
> 
> ```python
> with engine.connect() as conn:
>     insp = inspect(conn)
>     for tablename in insp.get_foreign_table_names():
>         Table(tablename, metadata, autoload_with=conn)
> ```

I think you may also pass the list of tables to reflect to `metadata.reflect` and it will do that.

I'm not sure if to fix this we should do something in particular or just document the use case above.
Otherwise we could kinda change how the metadata gets the list of columns and delegate to the inspect via a specific method. At this point we could introduce another flag that's "any_table_like" or similar that enables views and temp / fk / other strange things that 3rd party dialect may reflect?

**Comment by CaselIT** [comment]
> that wont work because it reconciles that list with what it gets from get_table_names()

indeed. new flag it is then I guess

**Comment by zzzeek** [comment]
so what are we here, "use case"?

**Comment by CaselIT** [comment]
yes probably. I'll try taking a look soonish

**Comment by CaselIT** [comment]
This seems somewhat problematic, at least judging by the fact that the only thing postgresql will get back for foreign tables are tables: no constraints (primary, check, fw, unique) no indexes and no comment

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Add get_any_table_names reflection method** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4734

**Comment by nmoreaud** [comment]
Do you plan to merge this PR?
I would be interested by this feature.

**Comment by CaselIT** [comment]
> @CaselIT can we take the individual methods here that are specific to PostgreSQL and add them to the PGInspector explicitly, so that people can build up commands like this from components that are PG specific for now?



I don't see what there would be to do for this. Users can already reflect them by using https://github.com/sqlalchemy/sqlalchemy/issues/9944#issuecomment-1593174953

**Comment by zzzeek** [comment]
oh when I looked at the patched that method seemed to be private.   

this feels pretty PG specific so i would want to just close this
